lib/HLSL/HLMatrixLowerPass.cpp

```diff
@@ -395,2445 +187,1310 @@ ModulePass *llvm::createHLMatrixLowerPass() { return new HLMatrixLowerPass(); }
+  // will temporarily create other instructions consuming/producing matrix types.
+  std::vector<AllocaInst*> MatAllocas;
+  std::vector<Instruction*> MatInsts;
+  getMatrixAllocasAndOtherInsts(Func, MatAllocas, MatInsts);
+
+  // First lower all allocas and take care of their GEP chains
+  for (AllocaInst* MatAlloca : MatAllocas) {
+    AllocaInst* LoweredAlloca = lowerAlloca(MatAlloca);
+    replaceAllVariableUses(MatAlloca, LoweredAlloca);
+    addToDeadInsts(MatAlloca);
+  }
+
+  // Now lower all other matrix instructions
+  for (Instruction *MatInst : MatInsts)
+    lowerInstruction(MatInst);
+
+  deleteDeadInsts();
+}
+
+void HLMatrixLowerPass::deleteDeadInsts() {
+  while (!m_deadInsts.empty()) {
+    Instruction *Inst = m_deadInsts.back();
+    m_deadInsts.pop_back();
+
+    DXASSERT_NOMSG(Inst->use_empty());
+    for (Value *Operand : Inst->operand_values()) {
+      Instruction *OperandInst = dyn_cast<Instruction>(Operand);
+      if (OperandInst && ++OperandInst->user_begin() == OperandInst->user_end()) {
+        // We were its only user, erase recursively.
+        // This will get rid of translation stubs:
+        // Original: MatConsumer(MatProducer)
+        // Producer lowered: MatConsumer(VecToMat(VecProducer)), MatProducer dead
+        // Consumer lowered: VecConsumer(VecProducer)), MatConsumer(VecToMat) dead
+        // Only by recursing on MatConsumer's operand do we delete the VecToMat stub.
+        DXASSERT_NOMSG(*OperandInst->user_begin() == Inst);
+        m_deadInsts.emplace_back(OperandInst);
       }
+    }
 
-      Type *FT = FunctionType::get(RetTy, params, false);
+    Inst->eraseFromParent();
+  }
+}
 
-      HLOpcodeGroup group = GetHLOpcodeGroupByName(CI->getCalledFunction());
-      unsigned opcode = GetHLOpcode(CI);
+// Find all instructions consuming or producing matrices,
+// directly or through pointers/arrays.
+void HLMatrixLowerPass::getMatrixAllocasAndOtherInsts(Function &Func,
+    std::vector<AllocaInst*> &MatAllocas, std::vector<Instruction*> &MatInsts){
+  for (BasicBlock &BasicBlock : Func) {
+    for (Instruction &Inst : BasicBlock) {
+      // Don't lower GEPs directly, we'll handle them as we lower the root pointer,
+      // typically a global variable or alloca.
+      if (isa<GetElementPtrInst>(&Inst)) continue;
 
-      Function *vecF = GetOrCreateHLFunction(*m_pModule, cast<FunctionType>(FT),
-                                             group, opcode);
+      if (AllocaInst *Alloca = dyn_cast<AllocaInst>(&Inst)) {
+        if (HLMatrixType::isMatrixOrPtrOrArrayPtr(Alloca->getType())) {
+          MatAllocas.emplace_back(Alloca);
+        }
+        continue;
+      }
+      
+      if (CallInst *Call = dyn_cast<CallInst>(&Inst)) {
+        // Lowering of global variables will have introduced
+        // vec-to-mat translation stubs, which we deal with indirectly,
+        // as we lower the instructions consuming them.
+        if (m_vecToMatStubs->contains(Call->getCalledFunction()))
```

Comment:
Could we have mat-to-vec stubs as well?  If so, wouldn't add them up as matrix consumers (MatInsts) below when we don't want to?

Reply:
hmm, I haven't hit the issue and I think it cannot happen because the only thing that happened before this is global lowering, which shouldn't have matrix operands (apart from the initializer which we lower). I'm only inserting vecToMat stubs on consumers of the matrix pointers, at the end of any GEP chain. It's a good idea to be more explicit here though, I'll see if I can make this an assert that `m_matToVecStubs` is still empty at that point.